collections/collections.go

```diff
@@ -0,0 +1,90 @@
+package collections
+
+import (
+	"errors"
+	"math"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+var (
+	// ErrNotFound is returned when the provided key is not present in the StorageProvider.
+	ErrNotFound = errors.New("collections: not found")
+	// ErrEncoding is returned when something fails during key or value encoding/decoding.
+	ErrEncoding = errors.New("collections: encoding error")
+)
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// Prefix defines a segregation namespace
+// for specific collections objects.
+type Prefix struct {
+	raw []byte // TODO(testinginprod): maybe add a humanized prefix?
+}
+
+// Bytes returns the raw Prefix bytes.
+func (n Prefix) Bytes() []byte { return n.raw }
+
+// NewPrefix returns a Prefix given the provided namespace identifier.
+// Prefixes of the same module must not start
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
```

Comment:
should we return an error instead?

Reply:
I thinkpanic is fine here, the switch case on `int` should really be a `uint8`, I put `int` because type inference works with `int`, meaning that you wouldn't need to do `collections.NewPrefix(uint8(1))` but just do `collections.NewPrefix(1)`. 

If the panic makes us unhappy I'd rather have developers explicitly write `NewPrefix(uint8(number))` than returning an error here. This function is just meant to return some union of types `[]byte,string,uint8` as prefix. Let me know.

Reply:
We can varint encode the prefix and let `int` be acceptable

Reply:
In general, I just think constructors should never panic -- they should return errors to their respective callers. But it's fine I suppose.